### PR TITLE
Luxi Lindsey - ride_share - Octos

### PR DESCRIPTION
**Pull request by Lindseyls**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I took a long time reviewing the code and the tests that were already written. Based on the design that was established by the previous programmer(s), I tried to keep my design the same style through out the new code and tests.
Describe any examples of composition that you encountered in this project, if any | There were lots of compositions throughout all the classes within the RideShare module. Such as, trips have passengers and drivers along with other attributes, then each passenger and diver have id, name, trips, and other attributes. When I wanted to create an instance of one class such as the driver class, I had to create an instance of the trip class first.
Describe a _nominal test_ that you wrote for this assignment. | I wrote probably my longest nominal test that I have ever written so far for the TripDispatcher#request_trip method. I was able to test that a new trip was created by passing a new passenger id into the parameter of the method. I checked that the passenger and driver information were correct and that what was expected was returned. 
Describe an _edge case test_ that you wrote for this assignment. | I wrote an edge case test for the TripDispatcher#request_trip method testing that the return value for time, cost, and rating are all nil since the trip is still in progress.
Describe a concept that you gained more clarity on as you worked on this assignment. | I definitely learned more about composition. Such as, learning how to read someone else's code and figuring out how the classes were composed of other classes. Also, I feel much more confident working on modules and classes compare to where I was a week ago. In addition, I still have a long ways to go before I can write clear and consist tests. I hope to work on that as much as possible... and how to DRY up the code.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -23,25 +22,55 @@ def initialize(input)
...
+      average = 0
+      finish_trip.each do |trip|
+        average += (total_revenue / (trip.duration / 3600.0))
+      end
+      return average.round(2)
+    end
+
+    def finish_trip
+      trips.reject {|trip| trip.end_time == nil}
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:73]: I love that you've broken this out as a separate helper method! This makes the other code in this file much easier to read.

File: lib/driver.rb [hunk 2]
```diff
@@ -23,25 +22,55 @@ def initialize(input)
...
+
+      total = subtotal * driver_takehome
+      return total.round(2)
+    end
+
+    def average_revenue
+      average = 0
+      finish_trip.each do |trip|
+        average += (total_revenue / (trip.duration / 3600.0))
+      end
```
Comment (droberts-sea) [inline lib/driver.rb:67]: I'm not sure this method does what you want. You add the total revenue (across all the trips) for _every_ trip, which will produce a much larger number than you want.

Something like this would be a correct implementation:

```ruby
def average_revenue
  total_time = 0
  finished_trip.each do |trip|
    total_time += trip.duration
  end
  return total_revenue / (total_time / 3600.0)
end
```

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -67,27 +68,82 @@ def load_trips
...
       end
 
-      trips
+      return trips
+    end
+
+    # this method goes through and finds the first driver with status that is :AVAILABLE
+    def available_driver
+      @drivers.find do |driver|
+        if driver.status == :AVAILABLE
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:109]: Good work making a helper method here. Having this logic separate would mean changing the way you select drivers (e.g. for wave 3) could be done without changing the `request_trip` method at all.

File: lib/trip_dispatcher.rb [hunk 4]
```diff
@@ -67,27 +68,82 @@ def load_trips
...
+      raise ArgumentError.new("There are no drivers available at this time")
+    end
+
+    def request_trip(passenger_id)
+
+      passenger = find_passenger(passenger_id)
+
+      if passenger.nil?
+        raise "Could not find passenger trip ID #{passenger_id}"
+      end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:122]: This error message talks about a trip ID, but interpolates the passenger ID. Which is it?

File: lib/driver.rb (hunk 1 above)
Comment (droberts-sea) [reply lib/driver.rb:73]: I am however not sure I agree with the name of this method. `finish_trip` sounds to me like a command: "Hey driver, finish your (most recent?) trip". Instead you might name it `finished_trips`, which sounds like it's asking for information (and also implies it'll be giving back an array instead of a single object).

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -76,5 +83,99 @@
...
+
+    it "correctly remove trip from trips if end time is nil" do
+      @driver.finish_trip.must_include @trip2
+      @driver.finish_trip.wont_include @trip
+    end
+
+    it "correctly only calculates trips that are not in-progress" do
+      @driver.average_rating.must_equal 5
+      @driver.total_revenue.must_equal 6.68
+      @driver.average_revenue.must_equal 6.68
```
Comment (droberts-sea) [inline specs/driver_spec.rb:178]: Organizationally, I would include these tests under the `describe` block for their respective methods.

File: specs/driver_spec.rb [hunk 6]
```diff
@@ -76,5 +83,99 @@
       driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
       driver.average_rating.must_equal 0
     end
-  end
-end
+  end # end of describe "average_rating method"
+
+  describe "total_revenue method" do
+    it "calculate the total revenue" do
+      trips = [
```
Comment (droberts-sea) [inline specs/driver_spec.rb:90]: One case I don't see your tests covering, for all the wave 1 methods, is what happens for a driver or passenger that has no trips at all?

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -88,5 +86,84 @@
...
+      trip = @dispatcher.request_trip(21)
+      second_driver = trip.driver
+      second_driver = @dispatcher.available_driver
+
+      second_driver.name.must_equal "Daryl Nitzsche"
+      second_driver.id.must_equal 3
+      second_driver.status.must_equal :AVAILABLE
+    end
+
+    # TODO: add a edge case here
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:114]: An example of an edge case would be, what happens if there are no available drivers? This is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.available_driver
  }.must_raise
end
```

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | yes
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | yes
Complete tests | yes
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | N/A
Appropriate use of Ruby's `Date` | N/A
Appropriate helper methods were made to help with complex logic | N/A
Complete tests | N/A
**Overall** | Good job overall! Your code is clear and easy to understand and your test coverage is on the right track. I have left a bunch of nitpicks below, and I found one serious logic bug, but in general I am quite pleased with this submission, and excited to see what you come up with on Hotel. Keep up the hard work!
